crates/flux-refineck/src/type_env.rs

```diff
@@ -508,8 +508,8 @@ impl BasicBlockEnvShape {
             (GenericArg::BaseTy(_), GenericArg::BaseTy(_)) => {
                 tracked_span_bug!("generic argument join for base types is not implemented")
             }
-            (GenericArg::Lifetime(re1), GenericArg::Lifetime(re2)) => {
-                debug_assert_eq!(re1, re2);
+            (GenericArg::Lifetime(re1), GenericArg::Lifetime(_re2)) => {
+                // CHECK: debug_assert_eq!(re1, re2);
```

Comment:
Just realize this is inside join and not subtyping. This feels bad. Can you minimize an example that needs this change?

Reply:
Yes, its rather tiny:

```rust
pub fn baz(v: &Vec<i32>) {
    for x in v {}
}
```

Reply:
Ok did some digging, this is indeed a bug somewhere else. 

When calling `Iter::next` we "downgrade" a pointer to a reference ([this line)](https://github.com/flux-rs/flux/blob/main/crates/flux-refineck/src/constraint_gen.rs#L265). 

So, we start with 

```
r: ptr(x), x: Iter<'?11, i32>
``` 
Then, just before the call, we have 

```
r: &mut Iter<?20, i32>, x: †Iter<'?20, i32>
```

i.e., the pointer `r` is now a mutable reference, and the type of `x` has been strongly updated (and blocked). The issue is that lifetimes should never be strongly updated. The call to `block_with` should update everything but lifetimes. That's what we do in assignments https://github.com/flux-rs/flux/blob/main/crates/flux-refineck/src/type_env.rs#L172.

The fix is a bit tricky because we are updating through path so we cannot do the same we do in `assign`.

Reply:
> The error also happens with iterating over RSet .. wonder why not with RVec??? - Ranjit.
> […](#)

The problem happens when we iterate `Vec` by reference. We don't have a by-reference iterator on `RVec`.